File: python/04_cases.py
"""Step 4: Full CRUD operations on Salesforce Cases.

Creates a temporary Account and Contact to link the Case to, then cleans up all three.

Run with:
    uv run python 04_cases.py
"""

from datetime import datetime
from auth import get_salesforce_connection


def main():
    sf = get_salesforce_connection()
    timestamp = datetime.now().strftime("%Y-%m-%dT%H:%M:%S")

    print("--- Case CRUD ---")

    # Create parent Account and Contact
    print("Setup: Creating parent Account and Contact...")
    account_result = sf.Account.create({"Name": f"Acme Corp {timestamp}"})
    account_id = account_result["id"]

    contact_result = sf.Contact.create({
        "LastName": f"Smith {timestamp}",
        "AccountId": account_id,
    })
    contact_id = contact_result["id"]
    print(f"  Account: {account_id}")
    print(f"  Contact: {contact_id}")

    # CREATE
    print("CREATE:")
    result = sf.Case.create({
        "Subject": f"Login button not working on mobile {timestamp}",   # Subject is required
        "Status": "New",
        "Priority": "High",
        "Origin": "Web",
        "AccountId": account_id,
        "ContactId": contact_id,
        "Description": "Customer reports the login button on the mobile app does not respond when tapped on iOS 17.",
    })
    case_id = result["id"]
    print(f"  Created Case: {case_id}")

    # READ
    print("READ:")
    case = sf.Case.get(case_id)
    print(f"  Case: Subject={case['Subject']}")
    print(f"  Status={case['Status']}, Priority={case['Priority']}")

    # UPDATE
    print("UPDATE:")
    sf.Case.update(case_id, {"Status": "Working"})
    case = sf.Case.get(case_id)
    print(f"  Updated Case: Status is now {case['Status']}")

    # DELETE
    print("DELETE:")
    sf.Case.delete(case_id)
    print(f"  Deleted Case: {case_id}")

    # Cleanup
    print("Cleanup: Deleting Contact and Account...")
    sf.Contact.delete(contact_id)
    sf.Account.delete(account_id)
    print("  Cleanup complete.")

    print("Done.")


if __name__ == "__main__":
    main()

File: javascript/04_cases.js
'use strict';
/**
 * Step 4: Full CRUD operations on Salesforce Cases.
 *
 * Creates a temporary Account and Contact to link the Case to, then cleans up all three.
 *
 * Run with:
 *   node 04_cases.js
 */

const { getConnection } = require('./auth');

async function main() {
    const conn = await getConnection();
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');

    console.log('--- Case CRUD ---');

    // Create parent Account and Contact
    console.log('Setup: Creating parent Account and Contact...');
    const accountResult = await conn.sobject('Account').create({ Name: `Acme Corp ${timestamp}` });
    const accountId = accountResult.id;

    const contactResult = await conn.sobject('Contact').create({
        LastName: `Smith ${timestamp}`,
        AccountId: accountId,
    });
    const contactId = contactResult.id;
    console.log(`  Account: ${accountId}`);
    console.log(`  Contact: ${contactId}`);

    // CREATE
    console.log('CREATE:');
    const createResult = await conn.sobject('Case').create({
        Subject: `Login button not working on mobile ${timestamp}`,   // Subject is required
        Status: 'New',
        Priority: 'High',
        Origin: 'Web',
        AccountId: accountId,
        ContactId: contactId,
        Description: 'Customer reports the login button on the mobile app does not respond when tapped on iOS 17.',
    });
    const caseId = createResult.id;
    console.log(`  Created Case: ${caseId}`);

    // READ
    console.log('READ:');
    const sfCase = await conn.sobject('Case').retrieve(caseId);
    console.log(`  Subject: ${sfCase.Subject}`);
    console.log(`  Status: ${sfCase.Status}, Priority: ${sfCase.Priority}`);

    // UPDATE
    console.log('UPDATE:');
    await conn.sobject('Case').update({ Id: caseId, Status: 'Working' });
    const updated = await conn.sobject('Case').retrieve(caseId);
    console.log(`  Updated Case: Status is now ${updated.Status}`);

    // DELETE
    console.log('DELETE:');
    await conn.sobject('Case').destroy(caseId);
    console.log(`  Deleted Case: ${caseId}`);

    // Cleanup
    console.log('Cleanup: Deleting Contact and Account...');
    await conn.sobject('Contact').destroy(contactId);
    await conn.sobject('Account').destroy(accountId);
    console.log('  Cleanup complete.');

    console.log('Done.');
}

main().catch(err => {
    console.error('Error:', err.message);
    process.exit(1);
});
